lib/customisation.nix

```diff
@@ -264,6 +266,24 @@ rec {
 
       outputsList = map outputToAttrListElement outputs;
     in commonAttrs // {
+      # unsafeDiscardOutputDependency:
+      # Not unsafe.
+      # Nix has two kinds of string contexts for derivation paths:
+      # - Regular store path dependency. It makes the `.drv` and the closure
+      #   of `.drvs` as well as all referenced input sources available.
+      # - DrvDeep, which causes all outputs of all derivations in the closure
+      #   to be added as dependencies. This behavior is rarely what you want,
+      #   unless, say, you're building a 25GB+ image containing everything to
+      #   rebuild each package in the closure of build dependencies for an
+      #   otherwise empty NixOS system. It'd be so big because it includes all
+      #   dependencies, all the way up to the bootstrap binaries.
+      # drvPath was traditionally a DrvDeep dependency, which makes it unsuitable
+      # for many use cases, including usage from a well-behaved Nix CLI.
+      # - `nix build foo.drvPath` _should_ mean: realise all those 25+ GB!
+      # - `nix build foo.derivation` does not realise anything, because it only
+      #   needs to instantiate a normal `.drv` closure, and that already constitutes
+      #   a valid store path.
+      derivation = assert condition; drv.derivation or (builtins.unsafeDiscardOutputDependency drv.drvPath);
```

Comment:
Is there a better name than `.derivation`? I feel like this can easily get confusing.

Didn't think much about this, but how about `.drvOnlyPath` for the new one and eventually `.drvClosurePath` for `.drvPath`?

Reply:
What would it be confused with?

I think you might be onto something though, because there could be more information about the derivation that we might want to put here. E.g. `{ path = "/nix/store/....drv"; drvAttrs = { ... }; inputDrvs = /* not available yet */; inputSrc = /* same problem */; }`.

Also I feel tempted to change `path` to `outPath`, which would be convenient, but also very wrong because there's nothing "out" about a `.drv`.
`derivation.drvPath` would be too confusing, with `drvPath` being this weird legacy thing.

Reply:
Hmm, is `derivation.path` still ok for the CLI use case? It's a little verbose.
Brevity is not essential here, because we'll have the lower-level alternative of `^..`, but that is somewhat of a drawback.

Reply:
> confusing.

Maybe it's because we are late to define _package attribute set_, which should take the place of some of the usage of _derivation_.

Good thing the Nix team decided to go ahead with [a definition of package attribute set](https://github.com/NixOS/nix/pull/9762/files) yesterday. It may take some time for that to seep through, until we have RFC 92 or multi-derivation packages as a forcing function.

Reply:
Confusing because there's already at least three notions of derivations:
1. The thing that `builtins.derivation` returns
2. The thing that `stdenv.mkDerivation` returns
3. The `.drv` paths and files in the Nix store

At least with `.drvPath` we distinguish that it's a _path_ to a derivation, which implies 3. But with `.derivation`, it's ambiguous just from the name. Even worse, it introduces a fourth notion of derivation, at least based on the name.

Reply:
`derivation.path` sounds better

I guess "path" generally might've been a bit confusing, since it's not actually a _path value type_ being returned, but rather a path in a _string_. But I'm not sure if that distinction is important here.

Reply:
Ok, so the plan is
1. `builtins.derivation` return value becomes legacy https://github.com/NixOS/nix/issues/9774
2. `stdenv.mkDerivation` returns a package attrset; not a derivation

As for (3): `.drv` is the serialised form of the accurate derivation concept. This is good, and I think the change to `derivation.path` clarifies the relation between the in-store derivation and the attrset that represents it.
A distinction between the real thing and its representation in a programming language happens all the time, even for things that are entirely virtual, such as files.

Reply:
This sounds pretty good to me!